### PR TITLE
Remove Device: Used to distribute all the bricks from device to other devices

### PR DESCRIPTION
**Pull request by MohamedAshiqrh**
Important:
We only support this feature for Replicate and Distribute Replicate Volume.
Refer: https://gluster.readthedocs.io/en/latest/Administrator%20Guide/Managing%20Volumes/#replace-brick

Removes all the brick from the device and start these bricks on other
devices based on allocator. Heal is triggered automatially for replicate volumes
on replace brick.

Allocate and create new brick to replace.
It stops the brick to be replaced, If it is not already down(kill the brick process).
Then gluster replace brick which will replace the brick with new one and also starts the heals.

If other nodes does not have sufficient storage then this command should fail.

Pair-Programmed-With: Mohamed Ashiq Liyazudeen mliyazud@redhat.com

Signed-off-by: Mohamed Ashiq Liyazudeen mliyazud@redhat.com
Signed-off-by: Raghavendra Talur rtalur@redhat.com

### CODE REVIEW

**Comment by MohamedAshiqrh** [comment]
Please DO NOT REVIEW. It is still breaking. Will update on the fixes and all things to do in the PR so that everyone is aware of the progress(and also can share their thoughts).

**Comment by lpabon** [comment]
@MohamedAshiqrh @raghavendra-talur I encourage you guys to try things out as PoCs, but on big changes like this you should first create a design document describing the changes and showing how it will deal with failures.  We can discuss that here in Github or over email.

File: apps/glusterfs/app.go [hunk 1]
```diff
@@ -291,6 +291,11 @@ func (a *App) SetRoutes(router *mux.Router) error {
 			Method:      "POST",
 			Pattern:     "/devices/{id:[A-Fa-f0-9]+}/state",
 			HandlerFunc: a.DeviceSetState},
+		rest.Route{
```
Comment (lpabon) [inline apps/glusterfs/app.go]: The model for device state changes is to use `/devices/{id..}/state`. It is then handled by the Node and Device state changes as seen [in node_entry.go](https://github.com/heketi/heketi/blob/master/apps/glusterfs/node_entry.go#L248), where the value that should be used to fail a node or device is `api.EntryStateFailed`.

File: apps/glusterfs/volume_entry_allocate.go [hunk 2]
```diff
@@ -63,6 +64,165 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+					return err
+				}
+				logger.Info("Remove node %v from allocator", node.Info.Id)
+				err = node.removeAllDisksFromRing(tx, allocator)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		defer func() {
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: This may not work since the db.Update() is called in a db.Update().  It should definitely be done another way.

File: apps/glusterfs/volume_entry_allocate.go [hunk 3]
```diff
@@ -63,6 +64,165 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
 	}
 }
 
+func (v *VolumeEntry) replaceBrickInVolume(db *bolt.DB, executor executors.Executor,
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: This function is _the_ most important part of the fix.  It needs to make sure that it picks a new brick from the ring that is on a _different_ zone than the ones already used and/or in the same zone if there is only one zone.  The algorithm used here should be discussed first.

File: apps/glusterfs/app.go (hunk 1 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/app.go]: @lpabon yeah agreed. so you are suggesting that we can use `/devices/{id:[A-Fa-f0-9]+}` instead of `/devices/{id:[A-Fa-f0-9]+}/remove` with method "POST". Then yes we can do it as we do not require a EntryState.

File: apps/glusterfs/volume_entry_allocate.go (hunk 3 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/volume_entry_allocate.go]: Sure.

File: apps/glusterfs/volume_entry_allocate.go (hunk 2 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/volume_entry_allocate.go]: done.

**Comment by MohamedAshiqrh** [comment]
@heketi/maintainers @heketi/dev The latest update of patch works as expected.
Zone is not yet considered. Will add the changes soon now replace brick is working as expected.

File: apps/glusterfs/volume_entry_allocate.go (hunk 3 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/volume_entry_allocate.go]: We can discuss more about this if you want in mail. 
http://lists.gluster.org/pipermail/heketi-devel/2017-February/000064.html

Comment (MohamedAshiqrh) [inline apps/glusterfs/volume_entry_allocate.go]: @lpabon ^^^

**Comment by raghavendra-talur** [comment]
@heketi/dev @heketi/maintainers @lpabon We were thinking about the problem of finding the right replacement for the brick. As heketi does not store the brick set information along with the volume or brick information, it is proving to be little tricky. We have two possible solutions:
1. Query the volume info in this function and determine on the equivalent of setList of VolumeCreate function. This seems doable and without breaking any backward compatibility.
2. Store the brick set information while we do a volume create. This however would mean that only those volumes that are created this version onwards are eligible for volume operations like replace brick etc. I know we can also do a one time query of gluster volumes and update the db with brick set info, but I am not inclined to do that.

If we do choose option 1, we don't really have to change anything in the allocator. Please let us know your thoughts.

Also, would you be open to taking this patch in as-is and considering the problem of "finding the right brick replacement" as a bug which we would fix in a week's time.

**Comment by MohamedAshiqrh** [comment]
@heketi/admin @heketi/dev I have added the volumeId to brickEntry and also made the changes to test. I feel the first option can be really helpful as We can support old volumes too. @lpabon Let us know what can be done.

**Comment by MohamedAshiqrh** [comment]
Also there was a small bug which was causing the http to not return as the DB was in transaction lock. The db was opened in view and replaceBrick was called. Inside replaceBrick we again opened the db with update so the transaction was hung in spurious cases. This is fixed in the updated commit.

**Comment by MohamedAshiqrh** [comment]
@heketi/dev @heketi/admin @lpabon Addressed all your comments, The Patch is clean and open for Review.

@raghavendra-talur and I worked on the logic of algorithm and Decisions were taken on the Algorithm based on the support for older volumes.

Prerequisites, Done:
We now added VolumeId in BrickEntry and VolumeInfo Executor call which will return Whole information of volume from gluster Itself(instead of saving the brick peer(brickset), we generate the brick peers from this information). 

How does this work:

For a Device to be remove.
First If the Device is Empty then Return ok to remove.
Else
Get the bricklist for bricks in device to be removed and its appropriate volumeEntrylist for bricks. 
Call Replace brick for a volume with the brickId.

In Replace Brick Logic:
1)First we Find the BrickSet(a set in which brick belongs, For Example in Distribute-Replicate 2x3 [A(A1,A2,A3), B(B1,B2,B3)], B2 is on set B) in which the brick to be replaced is present.
Reason to find this is we should not place the brick with another brick of same set(which will cause Quorum to be met if one node is down and also not a good design).
2) Call the allocator to give out devices for the same cluster.
3)Ignore the Device IF:
         a)Same Device to be removed
         b)Device belongs to same Node where one of the other bricks in Set is present
4) With above logic We can still use the logic of simpleAllocator ring to decide the brick placement with single Zone and Multiple zones.
5) On Failure returns Err and In case of NoSpaceError, We Respond Replacementnotfound.

Note:
Few basic tests added for New VolumeId for BrickEntry and all the failure based on executor.SimpleVolumeInfo change from executor.VolumeInfo has been fixed. 
Kept Device Remove modular so that can be used for Node Remove.

To Be Done: 
Tests to be Added.

Let us know If we missed anything. Open for Reviews.

**Comment by MohamedAshiqrh** [comment]
Looked into the Failing tests and fixed the Patch Hopefully the latest will pass the test.

**Comment by obnoxxx** [comment]
@centos-ci test this, please

File: executors/sshexec/volume.go [hunk 4]
```diff
@@ -228,3 +228,101 @@ func (s *SshExecutor) checkForSnapshots(host, volume string) error {
 
 	return nil
 }
+
+func (s *SshExecutor) VolumeInfo(host string, volume string) (*executors.SingleVolumeInfo, error) {
```
Comment (obnoxxx) [inline executors/sshexec/volume.go]: Could this possibly be added in a separate patch, in order to keep the patches a little smaller?

**Comment by raghavendra-talur** [comment]
@obnoxxx thanks for the review. Will address the comments soon.

File: apps/glusterfs/allocator_simple_test.go [hunk 5]
```diff
@@ -176,7 +176,7 @@ func TestSimpleAllocatorInitFromDbWithOfflineDevices(t *testing.T) {
 		// devices are added to the ring
 		node, err := cluster.NodeEntryFromClusterIndex(tx, 0)
 		tests.Assert(t, err == nil)
-		nodeId = node.Info.Id
+		//	nodeId := node.Info.Id
```
Comment (raghavendra-talur) [inline apps/glusterfs/allocator_simple_test.go]: @MohamedAshiqrh I need your help here. I don't remember doing this change, is it necessary?

File: executors/mockexec/mock.go [hunk 6]
```diff
@@ -9,9 +9,7 @@
 
 package mockexec
 
-import (
-	"github.com/heketi/heketi/executors"
-)
+import "github.com/heketi/heketi/executors"
```
Comment (raghavendra-talur) [inline executors/mockexec/mock.go]: Editor defaults, cleaning up.

File: executors/sshexec/volume.go (hunk 4 above)
Comment (raghavendra-talur) [inline executors/sshexec/volume.go]: Done.

File: apps/glusterfs/allocator_simple.go [hunk 7]
```diff
@@ -179,6 +180,10 @@ func (s *SimpleAllocator) GetNodes(clusterId, brickId string) (<-chan string,
 	// Get the list of devices for this brick id
 	devicelist, err := s.getDeviceList(clusterId, brickId)
 
+	for _, tester := range devicelist {
+		logger.Info("Allocator Device ID: %v, Brick ID: %v", tester.deviceId, brickId)
+	}
```
Comment (raghavendra-talur) [inline apps/glusterfs/allocator_simple.go]: Also, shall I move this to Debug level?

File: apps/glusterfs/allocator_simple_test.go (hunk 5 above)
Comment (raghavendra-talur) [inline apps/glusterfs/allocator_simple_test.go]: Ok, I think I got it. nodeId is unused, hence the removal?

Comment (MohamedAshiqrh) [inline apps/glusterfs/allocator_simple_test.go]: @raghavendra-talur It is not necessary. better to revert all the changes in the file as We are not depending on this.

File: apps/glusterfs/allocator_simple.go (hunk 7 above)
Comment (obnoxxx) [inline apps/glusterfs/allocator_simple.go]: Yeah, seems to call out for "Debug" ?

File: apps/glusterfs/allocator_simple_test.go (hunk 5 above)
Comment (obnoxxx) [inline apps/glusterfs/allocator_simple_test.go]: yeah, please remove unrelated changes. If there is a case for them, better do additional patches.

File: apps/glusterfs/allocator_simple.go [hunk 8]
```diff
@@ -10,8 +10,9 @@
 package glusterfs
 
 import (
-	"github.com/boltdb/bolt"
 	"sync"
+
+	"github.com/boltdb/bolt"
```
Comment (raghavendra-talur) [inline apps/glusterfs/allocator_simple.go]: Editor again, standard packages and github packages are separated by a space and standard packages should come first. Trying to revert this change.

Comment (obnoxxx) [inline apps/glusterfs/allocator_simple.go]: @raghavendra-talur if there is a case for this change -- make a separate patch (and maybe separate PR).

File: apps/glusterfs/brick_create.go [hunk 9]
```diff
@@ -57,6 +57,7 @@ func createDestroyConcurrently(db *bolt.DB,
 }
 
 func CreateBricks(db *bolt.DB, executor executors.Executor, brick_entries []*BrickEntry) error {
+	logger.Info("In Create Brick")
```
Comment (raghavendra-talur) [inline apps/glusterfs/brick_create.go]: done

File: apps/glusterfs/app_volume.go [hunk 10]
```diff
@@ -167,7 +167,7 @@ func (a *App) VolumeList(w http.ResponseWriter, r *http.Request) {
 
 	var list api.VolumeListResponse
 
-	// Get all the cluster ids from the DB
+	// Get all the volume ids from the DB
```
Comment (raghavendra-talur) [inline apps/glusterfs/app_volume.go]: done

File: apps/glusterfs/app.go (hunk 1 above)
Comment (raghavendra-talur) [inline apps/glusterfs/app.go]: @lpabon , @MohamedAshiqrh and I discussed this.

We liked that usage of api.* to indicate what to be done on the node/device instead of using another handler and REST URL. However, I don't feel "removed" fits in the hierarchy of states  "online" and "offline". Online and offline are mutually exclusive states and adding "removed" to the same hierarchy would corrupt it. A device would then could be in both "online" and "removed" state. Also, we feel that "failed" state is different than "removed". In remove we are just ensuring that all bricks from the device are moved away. It would be immediately available for new volumes.

In summary, we think, remove is an operation and not a change of state. Let us know what you think.

File: apps/glusterfs/app.go [hunk 11]
```diff
@@ -190,6 +197,15 @@ func (a *App) setLogLevel(level string) {
 	}
 }
 
+// Upgrade Path to update all the values for new API entries
+func (a *App) Upgrade(tx *bolt.Tx) error {
+	err := VolumeIdInBrickEntry(tx)
+	if err != nil {
+		logger.LogError("VolumeId add to BrickEntry failed")
+	}
```
Comment (raghavendra-talur) [inline apps/glusterfs/app.go]: need to return error here
change function name to "FillVolumeIdInBrickEntry"

File: apps/glusterfs/brick_entry.go [hunk 12]
```diff
@@ -12,6 +12,7 @@ package glusterfs
 import (
 	"bytes"
 	"encoding/gob"
+
```
Comment (raghavendra-talur) [inline apps/glusterfs/brick_entry.go]: This is recommended practice. One empty line before referencing non-standard packages.

**Comment by obnoxxx** [comment]
@centos-ci retest

**Comment by obnoxxx** [comment]
As far as I can tell, all of @lpabon 's requests have been addressed.
I am going to merge.
Awesome work, @MohamedAshiqrh and @raghavendra-talur ! 👍

File: apps/glusterfs/app.go (hunk 1 above)
Comment (lpabon) [inline apps/glusterfs/app.go]: What does "It would be immediately available for new volumes." mean?

Comment (raghavendra-talur) [inline apps/glusterfs/app.go]: @lpabon  A new volume create request might pick this device again.

Comment (lpabon) [inline apps/glusterfs/app.go]: @raghavendra-talur Why would we use the device again if the brick data was moved?  Why move it then?

File: apps/glusterfs/volume_entry_allocate.go [hunk 13]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+		if err != nil {
+			return err
+		}
+		oldBrickNode, err := NewNodeEntryFromId(tx, oldBrickEntry.Info.NodeId)
+		if err != nil {
+			return err
+		}
+
+		// Determine the setlist by getting data from Gluster
+		vinfo, err := executor.VolumeInfo(oldBrickNode.ManageHostName(), v.Info.Name)
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: Nicely done, but encapsulate line 116 to 138 in a function.

File: apps/glusterfs/volume_entry_allocate.go [hunk 14]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+		var slicestartindex int
+		var foundbrickset bool
+		var brick executors.Brick
+		setlist := make([]*BrickEntry, 0)
+		for slicestartindex = 0; slicestartindex <= len(vinfo.Bricks.Bricks)-v.Durability.BricksInSet(); slicestartindex = slicestartindex + v.Durability.BricksInSet() {
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: This needs a detailed explanation of what it is doing.

Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: What is Bricks.Bricks?

File: executors/executor.go [hunk 15]
```diff
@@ -69,5 +73,56 @@ type VolumeRequest struct {
...
+	DistCount       int      `xml:"distCount"`
+	StripeCount     int      `xml:"stripeCount"`
+	ReplicaCount    int      `xml:"replicaCount"`
+	ArbiterCount    int      `xml:"arbiterCount"`
+	DisperseCount   int      `xml:"disperseCount"`
+	RedundancyCount int      `xml:"redundancyCount"`
+	Type            int      `xml:"type"`
+	TypeStr         string   `xml:"typeStr"`
+	Transport       int      `xml:"transport"`
+	Bricks          Bricks
```
Comment (lpabon) [inline executors/executor.go]: Use another name for the structure

File: apps/glusterfs/volume_entry_allocate.go [hunk 16]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+			setlist = make([]*BrickEntry, 0)
+			for _, brick = range vinfo.Bricks.Bricks[slicestartindex : slicestartindex+v.Durability.BricksInSet()] {
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: Use a function which is more descriptive and easier to understand for v.info.Bricks.Bricks[...]...

File: apps/glusterfs/volume_entry_allocate.go (hunk 14 above)
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: This section would be easier if the bricks-in-set ids are saved in the db.  This would be the backup model.

Instead provide an "Upgrade()" call like the one above for volumes which do not have BricksInSet in the database, and fill in that information  Then in this function, just use the database.

File: apps/glusterfs/volume_entry_allocate.go [hunk 17]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+				brickentry, err := v.getEntryfromBrickName(tx, brick.Name)
+				if err != nil {
+					return err
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: We should log here `logger.LogError("Nice explanation: %v", err)`

File: apps/glusterfs/volume_entry_allocate.go [hunk 18]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+				}
+				if brickentry.Id() == oldBrickId {
+					foundbrickset = true
+				} else {
+					setlist = append(setlist, brickentry)
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: Nice method

File: apps/glusterfs/volume_entry_allocate.go [hunk 19]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+				}
+			}
+			if foundbrickset {
+				break
+			}
+		}
+		if !foundbrickset {
+			return err
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: logger.LogErr

File: apps/glusterfs/volume_entry_allocate.go [hunk 20]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+		}
+
+		for _, brickInSet := range setlist {
+			logger.Info("setlist is %v", brickInSet)
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: You can print it all using this: `logger.Info("Bricks in set %+v", setlist)`

File: apps/glusterfs/volume_entry_allocate.go [hunk 21]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+		}
+
+		logger.Info("Ask allocator for the list of devices")
+		// Check the ring for devices to place the brick
+		deviceCh, done, errc := allocator.GetNodes(v.Info.Cluster, newBrickId)
+		defer func() {
+			close(done)
+		}()
+
+		for deviceId := range deviceCh {
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: There is a lot of copied code here from the allocator.  Please redesign to minimize the copied code.

File: apps/glusterfs/volume_entry_allocate.go [hunk 22]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+		var newBrickId string
+		newId := false
+		for !newId {
+			newBrickId = utils.GenUUID()
+			if !utils.SortedStringHas(v.BricksIds(), newBrickId) {
+				newId = true
+			}
+		}
+
+		logger.Info("Ask allocator for the list of devices")
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: Remove this logging statement, it is confusing.  Or reword with information for the administrator

File: apps/glusterfs/volume_entry_allocate.go [hunk 23]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+			return nil, err
+		}
+		if brickname == fmt.Sprintf("%v:%v", nodeEntry.Info.Hostnames.Storage[0], brickentry.Info.Path) {
+			return brickentry, nil
+		}
+	}
+	return nil, ErrNotFound
+}
+
+func (v *VolumeEntry) replaceBrickInVolume(db *bolt.DB, executor executors.Executor,
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: This function is too large.  Please reorganize into one or two more functions.  Feel free to create a new xxx_replace.go file to place all these functions in.

File: apps/glusterfs/volume_entry_allocate.go [hunk 24]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+		}
+
+		for _, brickInSet := range setlist {
+			logger.Info("setlist is %v", brickInSet)
+		}
+
+		//Create an Id for new brick
+		var newBrickId string
+		newId := false
+		for !newId {
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: I don't think this is needed.  There none of this code in any other part of Heketi.  You can assume there will be no collisions.

File: apps/glusterfs/volume_entry_allocate.go [hunk 25]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+
+			// Do not allow a device from the same node to be
+			// in the set
+			deviceOk := true
+			for _, brickInSet := range setlist {
+				if brickInSet.Info.NodeId == newDeviceEntry.NodeId {
+					deviceOk = false
+				}
+			}
+
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: **Must have:**

Add a check here for Zone.  If the devices in the brick set not in the same zone, the new device must also not be in the same zone.

File: apps/glusterfs/volume_entry_allocate.go [hunk 26]
```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
...
+			logger.Info("Volume entry: Device ID: %v,Node ID: %v", newDeviceEntry.Id, newDeviceEntry.NodeId)
+			if !deviceOk {
+				continue
+			}
+			logger.Info("Got the device with id %v", newDeviceEntry.Id())
```
Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: Delete this log because we still do not know if the device will be accepted.

**Comment by lpabon** [comment]
First I want to add that this is really solid work.  The algorithm has one bug, but other than that it is very sound.  Unfortunately due to the number of issues shown, the code will be reverted.  Please, fix the above issues and resubmit the code.  This change is very important.

@MohamedAshiqrh @raghavendra-talur

File: apps/glusterfs/app.go [hunk 27]
```diff
@@ -290,6 +307,11 @@ func (a *App) SetRoutes(router *mux.Router) error {
 			Method:      "POST",
 			Pattern:     "/devices/{id:[A-Fa-f0-9]+}/state",
 			HandlerFunc: a.DeviceSetState},
+		rest.Route{
+			Name:        "DeviceRemove",
+			Method:      "POST",
+			Pattern:     "/devices/{id:[A-Fa-f0-9]+}/remove",
```
Comment (lpabon) [inline apps/glusterfs/app.go]: Another way it to use `DeviceDelete` to be able to delete a device and re-arrange all of its bricks.

File: apps/glusterfs/volume_entry_allocate.go (hunk 25 above)
Comment (raghavendra-talur) [inline apps/glusterfs/volume_entry_allocate.go]: @lpabon If I am not wrong, this is already done in allocator code. Even in allocBricks we don't check for zones and rely on allocator for that.

Comment (lpabon) [inline apps/glusterfs/volume_entry_allocate.go]: You have brought up a good point, but the allocator does not do that.  Which actually, is a bigger issue because this is also a problem with the normal allocation of bricks, where they could go into the same zone.

The allocator creates an array of the following form for a topology of three zones: [z1d0n0 z2d0n0, z3d0n0, z1d1n0, z2d1n0, z3d1n0, z1d0n1, z3d1n0, z3d1n0, and so on..].  So if z1d0n0 is the disk that needs to be replaced, it may choose z3d0n0.  But in the set for z1d0n0, it could have [z1d0n0, z2d1n0, z3d2n1], and if it chose z3d0n0 for replacing z1d0n0, now it has two drives in the same zone.  Instead it should make sure to go through the drives looking _first_ for drives _not_ in a zone already in the set.  Then if it does not find one because they have no space, then we may have to pick one in a zone which is already used as a fallback.

This sounds like it could be an issue also in the normal allocation, so solving it here we may use it as a fix for the normal allocation.

I'm not opposed to pushing this to its own fix after this if you want were we can concentrate the fix for both allocations in that one fix and provide the appropriate tests.  Up to you.

File: apps/glusterfs/volume_entry_allocate.go (hunk 14 above)
Comment (raghavendra-talur) [inline apps/glusterfs/volume_entry_allocate.go]: I am hesitating to populate db with this much of info because I want those changes to be properly designed. I request this change be taken in in current form. As it isn't a API change, it can be done in subsequent patches without breaking backward compatibility.

File: apps/glusterfs/app.go (hunk 1 above)
Comment (raghavendra-talur) [inline apps/glusterfs/app.go]: This is also available in pr #710

File: apps/glusterfs/app.go (hunk 27 above)
Comment (raghavendra-talur) [inline apps/glusterfs/app.go]: done

File: apps/glusterfs/app_device.go [hunk 28]
```diff
@@ -154,6 +154,56 @@ func (a *App) DeviceAdd(w http.ResponseWriter, r *http.Request) {
...
+			return logger.Err(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return
+	}
+	// Check and remove all bricks from device
+	logger.Info("Check and remove all bricks from device %v", device.Info.Id)
+	a.asyncManager.AsyncHttpRedirectFunc(w, r, func() (string, error) {
```
Comment (raghavendra-talur) [inline apps/glusterfs/app_device.go]: done

File: apps/glusterfs/app_device.go [hunk 29]
```diff
@@ -154,6 +154,56 @@ func (a *App) DeviceAdd(w http.ResponseWriter, r *http.Request) {
...
+		var err error
+
+		// Check if we can remove the device
+		if device.IsDeleteOk() {
+			logger.Info("Device is clean to delete or remove")
+			return "", nil
+		}
+
+		err = device.Remove(a.db, a.executor, a.allocator)
```
Comment (raghavendra-talur) [inline apps/glusterfs/app_device.go]: done with the functional change. Need more opinion on *naming*

File: apps/glusterfs/volume_entry_allocate.go (hunk 16 above)
Comment (raghavendra-talur) [inline apps/glusterfs/volume_entry_allocate.go]: I agree, changed the member name to be more descriptive. Will convert to function in refactor PR.

File: apps/glusterfs/volume_entry_allocate.go (hunk 21 above)
Comment (raghavendra-talur) [inline apps/glusterfs/volume_entry_allocate.go]: yes, will do.
